feat: new job to backup pins to s3

joshghent opened the pull request:
⚠️ Requires a new migration on the `psa_pins_request` table to add the new `backup_urls` column. ⚠️ 

### About this job
This new cron job which runs every 4 hours, grabs 10,000 pin requests, gets the car file and uploads it to s3.
It behaves in a similar way to [nftstorage/backup](https://github.com/nftstorage/backup).
It uses Dagula to grab the car file from the IPFS peer.

joshghent committed:
feat: added rough draft of the cron job to copy pins into s3

joshghent committed:
chore: add aws-sdk

joshghent committed:
feat: add cron

joshghent committed:
feat: added downloading and uploading of file correctly

joshghent committed:
chore: added required dependencies

joshghent committed:
feat: add new backup_urls column to the psa_pin_requests table

joshghent committed:
feat: finished job to backup files to s3

joshghent committed:
feat: added new test suite for pins backup

joshghent committed:
feat: add github action to run cron job

joshghent committed:
feat: added cron pins backup script

joshghent committed:
chore: removed todo

joshghent committed:
feat: added test mocks

joshghent committed:
chore: add mocks for export car

joshghent committed:
chore: update package-lock, fixed fetch reference

joshghent committed:
feat: swapped to a class tructure

joshghent committed:
chore: update the job to create a new backup class instance

joshghent committed:
feat: added more logging, updated query

joshghent committed:
fix: updated test mocks

joshghent committed:
fix: updated to make sql valid

flea89 committed:
Fix issues with updating rows

flea89 commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,239 @@
+  async s3Upload (s3, bucketName, bak) {
+    const key = `complete/${bak.contentCid}.car`
+    const region = await s3.config.region()
+    const url = new URL(`https://${bucketName}.s3.${region}.amazonaws.com/${key}`)
```
If I'm not mistaken, with the current implementation, given we can have different PSA_PIN_REQUESTS pointing to the same `CID`, we would end up uploading the same car multiple time to S3 isn't it?
I'm not quite aware if we need a particular folder structre, but assuming this one the correct one (and the same cid would always endup in the same path) we should check if already exists, and if yes, avoid uploading it?

Also, it seems from here that we always use content_cid (and not source_cid) for the key meaning different cid versions are not something we should be caring about?

joshghent replied:
Ok so, I've updated it so it's `complete/${bak.contentCid}/${bak.sourceCid}.car` for the key
And then added it so if that key already exists then we won't re-upload it.

flea89 replied:
Ace!

alanshaw replied:
The current convention is `complete/${bak.sourceCid}.car`

flea89 reviewed (changes requested):
This is getting together 🎉 
I did a really quick review pass and left some high-level comments for you.

Also noticed some types still need updating across the board.

joshghent commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,229 @@
+
+export default class Backup {
+  constructor (env) {
+    this.UPDATE_BACKUP_URL = 'UPDATE psa_pin_request SET backup_urls = $1 WHERE id = $2 AND content_cid = $3 RETURNING *'
```
I see why this was confusing. Updated :)

joshghent commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,229 @@
+      this.LIMIT
+    ])
+    if (!rows.length) return
+    const uploads = rows.filter(r => !r.url)
```
Updated

joshghent committed:
chore: remove redundant argument

joshghent committed:
chore: update variable names

joshghent commented on packages/cron/test/pins-backup.spec.js:
```diff
@@ -0,0 +1,198 @@
+  })
+
+  it.only('should attempt to backup pins to S3', async () => {
+    await backup.backupPins({ env, roPg, rwPg, cluster })
```
it used to! but I've removed it now :)

joshghent committed:
chore: update to use npm for the cron

joshghent commented on .github/workflows/cron-backup-pins.yaml:
```diff
@@ -0,0 +1,40 @@
+          STAGING_RO_PG_CONNECTION: ${{ secrets.STAGING_PG_CONNECTION }} # no replica for staging
+          PROD_PG_CONNECTION: ${{ secrets.PROD_PG_CONNECTION }}
+          PROD_RO_PG_CONNECTION: ${{ secrets.PROD_RO_PG_CONNECTION }}
+        run: yarn --cwd packages/cron start:pins:backup
```
yeah good point, I've updated to use npm

joshghent committed:
chore: updated type definition to accept a pool not client for pg

joshghent commented on packages/cron/src/bin/pins-backup.js:
```diff
@@ -0,0 +1,22 @@
+  const backup = new Backup(process.env)
+
+  try {
+    await backup.backupPins({ rwPg, roPg, cluster })
```
done

joshghent commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,229 @@
+import * as raw from 'multiformats/codecs/raw'
+import { Readable } from 'stream'
+
+export default class Backup {
```
yeah it's a good point, it was kind of just that "double-check". But yes this can be safely removed.

joshghent committed:
chore: remove commented code

joshghent commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,229 @@
+    this.SIZE_TIMEOUT = 1000 * 10 // timeout if we can't figure out the size in 10s
+    this.BLOCK_TIMEOUT = 1000 * 30 // timeout if we don't receive a block after 30s
+    this.REPORT_INTERVAL = 1000 * 60 // log download progress every minute
+    this.MAX_DAG_SIZE = 1024 * 1024 * 1024 * 32 // don't try to transfer a DAG that's bigger than 32GB
+    this.log = debug('backup:pins')
```
We don't... but it's a pattern that is found in other areas of the system. So might be worth flagging.

flea89 replied:
The comment is not pointing to the right line anymore, should be [line above](https://github.com/web3-storage/web3.storage/pull/1844/files#diff-4df8cd6f6a1194703527832b6e009d1515b1d8f708afe9a5af7c31f446f9aadeR25)[https://github.com/web3-storage/web3.storage/pull/1844/files#diff-4df8cd6f6a1194703527832b6e009d1515b1d8f708afe9a5af7c31f446f9aadeR25](url)

@alanshaw you might know more about this.

alanshaw replied:
My thinking was along the lines of that being the max in our ToS and wanting to have a cap at some point on the size of data we're willing to have uploaded/pinned. If you uploaded more than we've said is the max allowed then we shouldn't be obliged to store that.

flea89 replied:
But we don't have that limit for pins do we? I can't find the ToS for pins, but I might just be missing it?

Assuming there isn't one, there's a chance content bigger than that is stored there and users should expect it to be migrated?

flea89 replied:
I've commented out the size check part, because, unless we have strong reasons not to, we should move all the files to eipfs.

flea89 replied:
Add logging to keep track of enormous dags

flea89 replied:
I added 2 pieces of logging:
1. every time for whatever reason dag export fails we log the number of bytes read to that point
2. If a successful export is greater than 32TiB, we log it.

olizilla replied:
32TiB or 32 GiB? i think we want to know about pins >= 32 GiB

flea89 replied:
yeah, I actually added  32 GiB in the code 👍  (see [here](https://github.com/web3-storage/web3.storage/pull/1844/files#diff-4df8cd6f6a1194703527832b6e009d1515b1d8f708afe9a5af7c31f446f9aadeR34))

joshghent commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,229 @@
+    this.log = debug('backup:pins')
+    this.env = env
+    this.LIMIT = env.QUERY_LIMIT !== undefined ? env.QUERY_LIMIT : 10000
+    this.GET_PINNED_PINS_QUERY = `
```
Yes that is correct. There could be some extra work to add some de-duplication here. Not sure how common that case would be, but perhaps something for a V2.

joshghent replied:
I've updated the system so it will first check if the file is in S3 before uploading. But changes to the query could be made also.

joshghent committed:
feat: added way to not re-upload files we already have

joshghent committed:
chore: updated package lock

joshghent committed:
chore: updated package lock

joshghent commented:
Ok, so before this is merged it needs a the migration running which adds a new column `backup_urls` to the `psa_pins_request`.
This cron job will need verifying on staging and make sure that it successfully backs up the pins.
After that, it can be safely merged to production.

flea89 committed:
Make linter happier

alanshaw commented on .github/workflows/cron-backup-pins.yaml:
```diff
@@ -0,0 +1,40 @@
+
+on:
+  schedule:
+    - cron: '*/30 * * * *'
```
Why not just schedule for the max amount of time a job can run for 6h?

flea89 replied:
If I understand this correctly the job has 2 goals:
1. move historical pins to EIPFS
2. keep moving new psa requests to EIPFS until we move to pickup.

For **2** I guess it's ideal to keep moving stuff as promptly as possible (30 min make sense, even less than that?) while we know the first runs of the job will be super slow (since they will have to go through all the historical data).

Isn't a solution to satisfy both words to keep the `schedule` as is and set `concurrency` on the job?

alanshaw commented on packages/db/postgres/tables.sql:
```diff
@@ -307,11 +307,13 @@ CREATE TABLE IF NOT EXISTS psa_pin_request
   inserted_at     TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL,
-  updated_at      TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL
+  updated_at      TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL,
+  backup_urls     TEXT[]
```
Might be nice if this was `NOT NULL DEFAULT []` so you don't have to distinguish between null and empty.

alanshaw commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,248 @@
+     */
+    return async function * (source) {
+      for await (const candidate of source) {
+        yield { ...candidate, content: this.ipfsDagExport(ipfs, candidate.sourceCid, options) }
```
I think `this` will not be the expected object here.

alanshaw commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,248 @@
+        this.log(`received ${this.fmt(bytesReceived)} of ${formattedTotal} bytes`)
+      }, this.REPORT_INTERVAL)
+
+      for await (const chunk of ipfs.dagExport(cid, { timeout: this.BLOCK_TIMEOUT })) {
```
`dagExport` is not implemented on the IPFS Cluster client https://github.com/nftstorage/ipfs-cluster#api

alanshaw commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,248 @@
+  /**
+   * Export a CAR for the passed CID.
+   *
+   * @param {import('./ipfs-client').IpfsClient} ipfs
```
Does not exist!

alanshaw commented on packages/cron/test/pins-backup.spec.js:
```diff
@@ -0,0 +1,164 @@
+  DEBUG: 'backup:pins',
+  ENV: 'dev',
+  CLUSTER_API_URL: 'http://localhost:9094',
+  CLUSTER_IPFS_PROXY_API_URL: 'http://127.0.0.1:9095/api/v0/',
```
I think you probably want to connect directly to one of the nodes where the content is pinned. If you use the HTTP API via the DNS address you'll end up connecting to a random node and asking it to fetch the DAG (if it doesn't have it). There's no GC on these nodes so we want to avoid copying other nodes in the cluster where the data has not been allocated.

I'm also not certain the `api/v0/dag/export` route has been opened up.

In https://github.com/nftstorage/backup we use libp2p to connect to all nodes in the cluster and transfer via bitswap - only the nodes that are hosting the content will respond.

You could maybe do this with dagula nowadays if you look up which peer the content is hosted on.

alanshaw reviewed (changes requested):
You should check with @olizilla that the DB change is similar to what he's expecting to do.

olizilla committed:
feat: `w3 open <cid>` to open cid on w3s.link in browser (#1892)

I find myself having to add the https://w3s.link/ipfs prefix a lot. I
want a command to do it for me.

```bash
w3 open bafybeidluj5ub7okodgg5v6l4x3nytpivvcouuxgzuioa6vodg3xt2uqle

w3 open bafybeidluj5ub7okodgg5v6l4x3nytpivvcouuxgzuioa6vodg3xt2uqle/olizilla.png
```

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: put write to cluster behind a flag (#1785)

- Set `ENABLE_ADD_TO_CLUSTER` to send upload to ipfs-cluster after
sending response to user. The current behavior is to always try to write
user uploads to cluster. With this PR we disable that but leave a
feature flag to restore it if we need to.
- Adds LinkIndexer to decode and record links from each block as we
iterate through them. It allows us to set the DAG structure as
'Complete' where we know the upload contains all the blocks for a
complete DAG with no dangling links.
- Adds task to call to `LINKDEX_API` after the response is sent where
the DAG is `Partial`, to check if it is complete across other CARs under
the same upload.
- Where we have a complete CAR we set the status as "Pinned" early as
it's already in elastic-ipfs. This is debatable, as there is a lag until
it's discoverable, but I think it's reasonable as I want to push us
towards telling users when we safely have their stuff early.
- We write uploads to more than one place, by adding an R2Bucket binding
to our worker env called `carpark` where we write cars keyed by their
car CID.
- R2 only supports an [md5 integrity check on
put](https://developers.cloudflare.com/r2/runtime-apis/#r2putoptions),
so this PR re-uses the md5 lib that the aws-s3-sdk uses to generate that
and also stores it in the r2 metadata so we see what hash was used to
verify it later.
- CAR CIDs are generated as [w3up-cli does
it](https://github.com/web3-storage/w3up-cli/blob/b313c1221a0fb93a7d7cc3ced5fd1d4bce537a7a/src/utils.js#L46-L50)
with the [`0x202`
code](https://github.com/multiformats/multicodec/blob/955e61fe6221454c682ca9ba4901558bfc1bac96/table.csv#L130),
and added to the /car response body.
- miniflare is now available as a global in tests so we can make
assertions about [r2 bucket contents in
tests](https://miniflare.dev/storage/r2#manipulating-outside-workers),
which is nice.
- we make use of [miniflare's new fetch mocking
support](https://miniflare.dev/core/standards#mocking-outbound-fetch-requests)
to test calls to the linkdex-api happen when we expect them to.

_NOTE: We've been using ipfs-cluster as a DAG completeness check-ing
service, and we're using DAG completeness as a proxy for "we got all the
CARs and the DAG is stored". We need to do some reframing as it's a
feature of DAGs and CARs and IPFS that you could upload some `dab-json`
that links to the complete wikipedia archive. We want to inform you when
your blocks are safely stored, but we don't want to necessarily also
pull in all of wikipedia, nor do we want to avoid telling you your
blocks are pinned where they link to some other dataset._

Fixes: https://github.com/web3-storage/web3.storage/issues/1784

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

joshghent committed:
chore: update package lock

joshghent committed:
feat: pass through the query limit on workflow dispatch

joshghent committed:
feat: updated the backup urls column to default to an empty arr

joshghent committed:
chore: added dagula

joshghent committed:
feat: added default to empty array

joshghent committed:
feat: added default on the init data

joshghent committed:
feat: updated query to return correct data

joshghent committed:
chore: updated return types

joshghent committed:
chore: update package lock

joshghent committed:
chore: remove todo comment

flea89 committed:
Add fetch to cron pin-backup

flea89 committed:
Update types

flea89 committed:
Add Minio to cron

flea89 committed:
Add getS3 to cron utils

flea89 committed:
update tests to be more e2endy

flea89 committed:
Comment out size limiting

flea89 committed:
more type fixes

flea89 committed:
Use s3 v3 and updats to backup updates

flea89 commented on packages/api/src/car.js:
```diff
@@ -6,6 +6,7 @@ import { toString, equals } from 'uint8arrays'
 import { LinkIndexer } from 'linkdex'
```
Not sure where these changes are coming from? I wonder if this is coming from a bad merge.
Pretty sure all changes in API package needs to be reverted to whatever we have in main

flea89 replied:
Reverted changest to the API

flea89 committed:
tidy tests up

flea89 committed:
Set concurrency to avoid multiple jobs running together

flea89 committed:
Use ipfs node id rather than cluster one

flea89 committed:
Create a car file for upload and test it holds the expected content

flea89 committed:
Add abort controller

flea89 committed:
Merge branch 'main' into feat/794-copy-pins-to-eips

flea89 committed:
Align api package with main

flea89 committed:
Improve logging and more tests

flea89 committed:
Improve testing and error handling

flea89 committed:
Remove only from test

flea89 committed:
Update logging

flea89 committed:
Update node modules

flea89 committed:
Update setup node action

flea89 committed:
Fix dependecies

flea89 committed:
Increase cron timeout

flea89 commented:
@olizilla can you please give this a thorough review 🙏 
Let me know if you want to take it from here, or you will need more support from my end as well.

flea89 commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,301 @@
+      let reportInterval
+      const libp2p = await getLibp2p()
+      try {
+        const dagula = await Dagula.fromNetwork(libp2p, { peer })
```
Cache instances for same peer location

flea89 replied:
Done.

flea89 commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,301 @@
+        throw (err)
+      } finally {
+        if (bytesReceived > this.MAX_UPLOAD_DAG_SIZE) {
+          this.log(`⚠️ CID: ${cid} dag is greater than ${this.fmt(this.MAX_UPLOAD_DAG_SIZE)}`)
```
Add a per batch summary:
- failed cids (and their size)
- Successful (bigger than MAX_UPLOAD_DAG_SIZE)

Remove all unnecessary per cids logging, log just errors by default.

flea89 replied:
Done!

Default logging (`DEBUG=backupPins:log` ) is quite succinct now, while the job can be run manually passing a more verbose `DEBUG=backupPins:*` through workflow inputs.

Example of default logging:
``` 
❯ DEBUG=backupPins:log npm test --workspace=packages/cron
```
![Screenshot 2022-10-31 at 16 24 12](https://user-images.githubusercontent.com/3163684/199058522-e6ad8729-246e-4466-9721-78958f1324af.png)

``` 
❯ DEBUG=backupPins:log npm test --workspace=packages/cron
```
![Screenshot 2022-10-31 at 16 26 17](https://user-images.githubusercontent.com/3163684/199058942-1fdae1dd-5ddc-4564-afcd-22f5be054e8c.png)

flea89 commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,301 @@
+   * @returns {Promise<number | undefined>}
+   */
+
+  // Given for PIN requests we never limited files size we shouldn't check this. ie.
```
Delete stale code

flea89 commented on packages/cron/src/jobs/pins-backup.js:
```diff
@@ -0,0 +1,301 @@
+            Bucket: bucketName,
+            Key: key,
+            Body: bak.content,
+            Metadata: { structure: 'Complete' }
```
Look into sending checksum for file, reject the upload if the bytes of car don't match the cid

flea89 replied:
I wonder if this is required.
Looking at the headers sent by the client 
```
  'content-type': 'application/xml',
  'content-length': '11985',
  Expect: '100-continue',
  host: '127.0.0.1',
  'x-amz-user-agent': 'aws-sdk-js/3.53.1',
  'user-agent': 'aws-sdk-js/3.53.1 os/darwin/21.6.0 lang/js md/nodejs/16.14.0 md/crt-avail api/s3/3.53.1',
  'amz-sdk-invocation-id': '25110079-acaf-425f-8933-3527fd8366c7',
  'amz-sdk-request': 'attempt=1; max=3',
  'x-amz-date': '20221031T122520Z',
  'x-amz-content-sha256': 'ebd8a0f42b66a7756aaee73e6275d918143525f125137b584e6b079b364a6b5f',
  authorization: 'AWS4-HMAC-SHA256 Credential=minioadmin/20221031/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-user-agent, Signature=90453c633c07234480f3319eb3c1b058d25b39a077eb3653063165c5bc137722'
}
```

you can see it sends `x-amz-content-sha256` which is the sha256 of the payload and implies the payload is signed (see [docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html).

If every chunk is hashed and verified I don't think we need the overall one? Or am I missing something?

flea89 replied:
I wonder if this is required.
Looking at the headers sent by the client 
```
  'content-type': 'application/xml',
  'content-length': '11985',
  Expect: '100-continue',
  host: '127.0.0.1',
  'x-amz-user-agent': 'aws-sdk-js/3.53.1',
  'user-agent': 'aws-sdk-js/3.53.1 os/darwin/21.6.0 lang/js md/nodejs/16.14.0 md/crt-avail api/s3/3.53.1',
  'amz-sdk-invocation-id': '25110079-acaf-425f-8933-3527fd8366c7',
  'amz-sdk-request': 'attempt=1; max=3',
  'x-amz-date': '20221031T122520Z',
  'x-amz-content-sha256': 'ebd8a0f42b66a7756aaee73e6275d918143525f125137b584e6b079b364a6b5f',
  authorization: 'AWS4-HMAC-SHA256 Credential=minioadmin/20221031/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-user-agent, Signature=90453c633c07234480f3319eb3c1b058d25b39a077eb3653063165c5bc137722'
}
```

you can see it sends `x-amz-content-sha256` which is the sha256 of the payload and implies the payload is signed (see [docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html).

If every chunk is hashed and verified I don't think we need the overall one? Or am I missing something?

flea89 replied:
cc @olizilla

flea89 committed:
Improve logging and remove stale code

flea89 committed:
Cache instances of Dagula and use 1 libp2p

flea89 committed:
Better instanciation of resources

flea89 committed:
Default to quiet logging, but allow for more verbose when running manually

flea89 committed:
Update cron help text
